### PR TITLE
Implement the tools to solve the NOCI problem.

### LINKED ISSUE #998: implement the components to solve the NOCI problem
To solve the NOCI problem we need to implement components, conform to the structure of GQCP.

- [ ] A NOCI environment
- [ ] A NOCI QC Method
- [ ] A NOCI QC Model
- [ ] Make sure QC Structure works with any NOCI model

### PR DESCRIPTION
**Pull request by xdvriend**
**Short description**

Since NOCI is inherently a `CI` method, I wanted to mimic the implementation as close as possible. This explains the implemented parts in this PR.

**Related issues**

- closes #998 

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files
- [x] Implement NOCI QCMethod.
- [x] Implement NOCIEnvironment.
- [x] implement NOCI QCModel.
- [x] Implement GeneralizedEigensolver to mimic FCI structure.
- [x] Implement 1- ~and 2~-DM in the QCModel.
- [x] Test the implementations.

EDIT: 2DM will not be implemented as of yet. The formula I wanted to use (Eq. 54 in [Burton 2021](https://doi.org/10.1063/5.0045442) describes the `transition2DM` of the biorthogonal basis. What the 2DM of the NOCI model looks like must still be figured out.)

### CODE REVIEW

File: gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#include "Mathematical/Optimization/Eigenproblem/GeneralizedEigenproblemEnvironment.hpp"
+
+#include <Eigen/Dense>
+
+
+namespace GQCP {
+
+
+/**
+ *  A step that performs a dense diagonalization.
```
Comment (xdvriend) [inline gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp]:
```suggestion
 *  A step that performs a dense generalized diagonalization.
```

File: gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public Step<GeneralizedEigenproblemEnvironment<_Scalar>> {
+public:
+    // The scalar type of the matrix elements: real or complex.
+    using Scalar = _Scalar;
+
+
+public:
+    /*
+     *  PUBLIC OVERRIDDEN METHODS
+     */
```
Comment (xdvriend) [inline gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp]:
```suggestion
    /**
     *  MARK: Overridden methods
     */
```

File: gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     *  @return a textual description of this algorithmic step
```
Comment (xdvriend) [inline gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp]:
```suggestion
     *  @return A textual description of this algorithmic step.
```

File: gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    std::string description() const override {
+        return "Read the matrix from the environment, diagonalize it and write the number of requested eigenpairs to it.";
+    }
+
+
+    /**
+     *  Read the matrix from the environment, diagonalize it and write the number of requested eigenpairs to it.
+     *
+     *  @param environment              the environment that this step can read from and write to
+     */
```
Comment (xdvriend) [inline gqcp/include/Mathematical/Optimization/Eigenproblem/GeneralizedDenseDiagonalization.hpp]:
```suggestion
    /**
     *  Read the matrix from the environment, diagonalize it and write the number of requested eigenpairs to it.
     *
     *  @param environment              The environment that this step can read from and write to.
     */
```

File: gqcp/include/QCMethod/NOCI/NOCI.hpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+#include <memory>
+
+
+namespace GQCP {
+namespace QCMethod {
+
+
+/**
+ *  The configuration interaction quantum chemical method.
```
Comment (xdvriend) [inline gqcp/include/QCMethod/NOCI/NOCI.hpp]:
```suggestion
 *  The non-orthogonal configuration interaction quantum chemical method.
```

File: gqcp/include/QCMethod/NOCI/NOCI.hpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ *  @tparam _Scalar                       The scalar type of the expansion coefficients: real or complex.
+ *  @tparam _NonOrthogonalBasis           The type of ONV basis.
```
Comment (xdvriend) [inline gqcp/include/QCMethod/NOCI/NOCI.hpp]:
```suggestion
 *  @tparam _NonOrthogonalBasis           The type of non-orthogonal state basis.
```

File: gqcp/include/QCMethod/NOCI/NOCI.hpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+     *
+     *  @param solver               The solver that will try to optimize the parameters.
+     */
+    template <typename Solver>
+    QCStructure<NOCIExpansion<Scalar, NonOrthogonalBasis>, Scalar> optimize(Solver& solver, GeneralizedEigenproblemEnvironment<Scalar>& environment) const {
+
+        // The NOCI method's responsibility is to try to optimize the parameters of its method, given a solver and associated environment.
+        solver.perform(environment);
+
+
```
Comment (xdvriend) [inline gqcp/include/QCMethod/NOCI/NOCI.hpp]:
```suggestion
```

File: gqcp/include/QCModel/NOCI/NOCIExpansion.hpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+            }
+        }
+        return Orbital1DM<double> {D};
+    }
+
+
+    /**
+     *  Calculate the orbital one-electron density matrix for a complex valued expansion in a restricted non-orthogonal basis.
+     *
+     *  @return The generalized one-electron density matrix.
```
Comment (xdvriend) [inline gqcp/include/QCModel/NOCI/NOCIExpansion.hpp]:
```suggestion
     *  @return The orbital one-electron density matrix.
```

File: gqcp/include/QCModel/NOCI/NOCIExpansion.hpp [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+
+    /**
+     * MARK: Density maytrices for `UNonOrthogonalStateBases`
+     */
+
+    /**
+     *  Calculate the spin-resolved one-electron density matrix for a real valued expansion in a unrestricted non-orthogonal basis.
+     *
+     *  @return The orbital one-electron density matrix.
```
Comment (xdvriend) [inline gqcp/include/QCModel/NOCI/NOCIExpansion.hpp]:
```suggestion
     *  @return The spin-resolved one-electron density matrix.
```

File: gqcp/include/QCModel/NOCI/NOCIExpansion.hpp [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+            }
+        }
+        return SpinResolved1DM<double> {D_a, D_b};
+    }
+
+
+    /**
+     *  Calculate the spin-resolved one-electron density matrix for a complex valued expansion in a unrestricted non-orthogonal basis.
+     *
+     *  @return The generalized one-electron density matrix.
```
Comment (xdvriend) [inline gqcp/include/QCModel/NOCI/NOCIExpansion.hpp]:
```suggestion
     *  @return The spin-resolved one-electron density matrix.
```

File: gqcp/include/QCModel/NOCI/NOCIExpansion.hpp [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+
+
+/**
+ *  A class that represents an expansion inside a non-orthogonal basis.
+ *
+ *  @tparam _Scalar                       The scalar type of the expansion coefficients: real or complex.
+ *  @tparam _NonOrthogonalBasis           The type of non-orthogonal basis.
+ */
+template <typename _Scalar, typename _NonOrthogonalBasis>
+class NOCIExpansion {
```
Comment (dariavh) [inline gqcp/include/QCModel/NOCI/NOCIExpansion.hpp:48]: Is there a possibility that you could inherit from the LinearExpansion class?

**Review by dariavh — APPROVED** [review]

File: gqcp/include/QCModel/NOCI/NOCIExpansion.hpp (hunk 11 above)
Comment (xdvriend) [reply gqcp/include/QCModel/NOCI/NOCIExpansion.hpp:48]: All calculations of density matrices and the basis on which it is based are completely different, so I don't really think that's a possibility, unfortunately 😅

**Pull request merged**